File: LeetCodePractice/src/designDataStructure/MinStack.java
package designDataStructure;

import java.util.*;

/*
#155

Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

push(x) -- Push element x onto stack.
pop() -- Removes the element on top of the stack.
top() -- Get the top element.
getMin() -- Retrieve the minimum element in the stack.


Example 1:

Input
["MinStack","push","push","push","getMin","pop","top","getMin"]
[[],[-2],[0],[-3],[],[],[],[]]

Output
[null,null,null,null,-3,null,0,-2]

Explanation
MinStack maxStack = new MinStack();
maxStack.push(-2);
maxStack.push(0);
maxStack.push(-3);
maxStack.getMin(); // return -3
maxStack.pop();
maxStack.top();    // return 0
maxStack.getMin(); // return -2


Constraints:

Methods pop, top and getMin operations will always be called on non-empty stacks.

analysis:
min stack is decreasing order
 */

public class MinStack {
	Stack<Integer> realStack = new Stack<>();
	Stack<Integer> minStack = new Stack<>();
	
    public void push(int x) {
	    if(minStack.isEmpty() || minStack.peek() >= x){
	    	minStack.push(x);
	    }
        realStack.push(x);
    }

   public void pop() {
		if (minStack.peek().equals(realStack.peek())) {
			minStack.pop();
		}
		realStack.pop();
	}

    public int top() {
        return realStack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }
}
/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: LeetCodePython/MinStack.py
"""
Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

Implement the MinStack class:

MinStack() initializes the stack object.
void push(int val) pushes the element val onto the stack.
void pop() removes the element on the top of the stack.
int top() gets the top element of the stack.
int getMin() retrieves the minimum element in the stack.


Example 1:

Input
["MinStack","push","push","push","getMin","pop","top","getMin"]
[[],[-2],[0],[-3],[],[],[],[]]

Output
[null,null,null,null,-3,null,0,-2]

Explanation
MinStack minStack = new MinStack();
minStack.push(-2);
minStack.push(0);
minStack.push(-3);
minStack.getMin(); // return -3
minStack.pop();
minStack.top();    // return 0
minStack.getMin(); // return -2


Constraints:
-231 <= val <= 231 - 1
Methods pop, top and getMin operations will always be called on non-empty stacks.
At most 3 * 104 calls will be made to push, pop, top, and getMin.

hint:
1 Consider each node in the stack having a minimum value.
"""

class MinStack:

    def __init__(self):
        self.st = []

    def push(self, val: int) -> None:
        least = self.getMin()
        if least is None or val < least:
            least = val
        self.st.append((val, least))

    def pop(self) -> None:
        self.st.pop()

    def top(self) -> int:
        if not self.st:
            return None
        return self.st[-1][0]

    def getMin(self) -> int:
        if not self.st:
            return None
        return self.st[-1][1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()